error: could not compile

```
error[E0557]: feature has been removed
 --> /Users/shogitai/.cargo/registry/src/github.com-1ecc6299db9ec823/state-0.2.2/src/lib.rs:1:12
  |
1 | #![feature(const_fn)]
  |            ^^^^^^^^ feature has been removed
  |
```

I run `cargo build` but I give the above error.

----
>  `rustc 1.62.0 (a8314ef7d 2022-06-27)` 
> `cargo 1.62.0 (a748cf5a3 2022-06-08)`

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "ots-viewer"
-version = "0.1.0"
+version = "0.2.0"
 authors = ["Andrew Poelstra <apoelstra@wpsoftware.net>"]
 
 [[bin]]
@@ -9,16 +9,17 @@ path = "src/main.rs"
 
 [dependencies]
 bitcoin = "0.9"
+rocket-multipart-form-data = "0.9"
 rust-crypto = "0.2"
-multipart = "0.11"
 opentimestamps = "0.1"
-rocket = "0.2"
-rocket_codegen = "0.2"
-serde = "0.9"
-serde_derive = "0.9"
+rocket = "0.4"
 
 [dependencies.rocket_contrib]
-version = "0.2"
+version = "0.4"
 default-features = false
 features = ["handlebars_templates"]
+
+[dependencies.serde]
+version = "1"
+features = [ "derive" ]
 

--- a/src/main.rs
+++ b/src/main.rs
@@ -23,20 +23,17 @@
 #![deny(non_snake_case)]
 #![deny(unused_mut)]
 
-#![feature(plugin)]
-#![plugin(rocket_codegen)]
+#![feature(decl_macro)]
 
 extern crate bitcoin;
 extern crate crypto;
-extern crate multipart;
+extern crate rocket_multipart_form_data;
 extern crate opentimestamps as ots;
 extern crate rocket_contrib;
-extern crate rocket;
-extern crate serde;
-#[macro_use] extern crate serde_derive;
+#[macro_use] extern crate rocket;
+#[macro_use] extern crate serde;
 
 use std::collections::HashMap;
-use std::borrow::Cow;
 use std::fs;
 use std::path::{Path, PathBuf};
 
@@ -44,23 +41,16 @@ use bitcoin::blockdata::transaction::Transaction;
 use bitcoin::network::serialize::{deserialize, BitcoinHash};
 use crypto::digest::Digest;
 use crypto::sha2::Sha256;
+use rocket_multipart_form_data::{MultipartFormDataOptions, MultipartFormData, MultipartFormDataField};
 use ots::attestation::Attestation;
 use ots::timestamp::{Step, StepData};
 use ots::op::Op;
 use ots::hex::Hexed;
-use rocket::http::ascii::UncasedAscii;
+use rocket::Data;
 use rocket::http::ContentType;
 use rocket::response::content;
 use rocket::response::{Redirect, NamedFile};
-use rocket_contrib::Template;
-
-pub mod multipart_stream;
-
-const OCTET_STREAM: ContentType = ContentType {
-    ttype: UncasedAscii { string: Cow::Borrowed("application") },
-    subtype: UncasedAscii { string: Cow::Borrowed("octet-stream") },
-    params: None
-};
+use rocket_contrib::templates::Template;
 
 #[derive(Debug, Serialize)]
 struct DisplayedStep {
@@ -193,8 +183,9 @@ fn view(file: PathBuf) -> Template {
 // Download
 #[get("/download/<file..>")]
 fn download(file: PathBuf) -> Option<content::Content<NamedFile>> {
+    let octet_stream: ContentType = ContentType::new("application", "octet-stream");
     if let Ok(nf) = NamedFile::open(Path::new("cache/").join(file)) {
-        Some(content::Content(OCTET_STREAM, nf))
+        Some(content::Content(octet_stream, nf))
     } else {
         None
     }
@@ -220,8 +211,27 @@ fn doc_id(dtf: &ots::DetachedTimestampFile) -> String {
 
 // Upload handler
 #[post("/upload", data="<ots>")]
-fn upload(ots: multipart_stream::MultipartStream) -> Redirect {
-    match ots::DetachedTimestampFile::from_reader(ots.stream) {
+fn upload(content_type: &ContentType, ots: Data) -> Redirect {
+    let options = MultipartFormDataOptions::with_multipart_form_data_fields(
+        vec![MultipartFormDataField::file("file")]
+    );
+    let multipart_form_data = MultipartFormData::parse(content_type, ots, options).unwrap();
+    let filepath = match multipart_form_data.files.get("file") {
+        Some(ref file) => &file[0].path,
+        None => {
+            println!("No file provided.");
+            return Redirect::to("/");
+        }
+    };
+    let fh = match fs::File::open(filepath) {
+        Ok(fh) => fh,
+        Err(e) => {
+            println!("Failed to open uploaded file: {}", e);
+            return Redirect::to("/");
+        }
+    };
+
+    match ots::DetachedTimestampFile::from_reader(fh) {
         Ok(dtf) => {
             let id = doc_id(&dtf);
             match fs::File::create(Path::new("cache/").join(&id)) {
@@ -230,7 +240,7 @@ fn upload(ots: multipart_stream::MultipartStream) -> Redirect {
                         println!("Filed to write timestamp: {}", e);
                         Redirect::to("/")
                     } else {
-                        Redirect::to(&format!("/view/{}", id))
+                        Redirect::to(format!("/view/{}", id))
                     }
                 }
                 Err(e) => {

--- a/src/multipart_stream.rs
+++ b/src/multipart_stream.rs
@@ -28,11 +28,13 @@ const SIZE_LIMIT: u64 = 32768;
 
 /// A wrapper around a multipart stream
 pub struct MultipartStream {
-    pub stream: Box<Read>
+    pub stream: Box<dyn Read>
 }
 
-impl FromData for MultipartStream {
+impl<'a> FromData<'a> for MultipartStream {
     type Error = &'static str;
+    type Owned = &'static str;
+    type Borrowed = &'static str;
 
     fn from_data(request: &Request, data: Data) -> data::Outcome<Self, Self::Error> {
         let ct = match request.headers().get_one("Content-Type") {

